[25.0] Fix deferred datasets in multiple dataset parameters.

jmchilton opened the pull request:
Alternative to #20605. Fixes https://github.com/galaxyproject/galaxy/issues/20270 with a test to exercise the problem. The problem is I was using the input_datasets dict that completely de-structures array inputs and such. Sticking closer to the tool state and reasoning about arrays of inputs properly seems to resolve the problem.

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

jmchilton committed:
Migrate some tool-on-deferred tests to newer style execute tests.

jmchilton committed:
Fix deferred datasets in multi-data parameters.

mvdbeek commented:
yeah, that did end up in error again: https://github.com/galaxyproject/galaxy/actions/runs/16226342461/job/45991121442?pr=20650#step:16:1632

jmchilton committed:
Fix using multiple data inputs with deferred datasets & Pulsar.

We really have to eliminate that input_datasets dictionary - this is very hacky. Maybe we can take a pass at rewriting Pulsar staging after we have stronger typing in the tool state branch for this stuff. Now we're putting multiple copies of an HDA into the deferred object structure - so I am now caching them during materialization so we don't materialize them more than once - this was probably needed anyway regardless since an input can be supplied to a tool more than once.

mvdbeek commented:
Thanks @jmchilton!
